mcmetadata/content.py

```diff
@@ -214,10 +217,12 @@ def extract(self, url: str, html_text: str, include_metadata: bool = False):
         for t in text:
             if t.parent.name not in self.REMOVE_LIST:
                 output += '{} '.format(t)
+        canonical_url = soup.find('link', rel='canonical')['href'] if soup.find('link', rel='canonical') else None
```

Comment:
While this code rarely gets called, I wonder:
Is `soup.find` expensive (does it do a full document tree traversal)?
I tend to doubt Python will do a common sub-expression elimination
(since I don't _think_ it can know if the call has side-effects).

If one was DETERMINED to do it in one statement, I suppose
`canonical_url = canonical_link['href'] if (canonical_link := soup.find('link', rel='canonical')) else None`
might do the trick, tho personally I find the ordering of the trinary operator in Python unsettling
(I grew up as a C programmer, where `condition ? true_expr : false_expr` seems clearer), and I'd probably
code it in three lines:

Reply:
```
can_url = None
if (can_link := soup.find("link", rel="canonical")):
    can_url = can_link.get("href")
```

Note: we've enabled "pre-commit" hooks in many repositories, and the `black` settings prefer (change) all single quotes to double quotes!

Reply:
In the worst-case scenario, if there is no canonical link tag, `soup.find` will perform a full tree traversal, as it only short-circuits upon finding the first match. I see how expensive this can be. Since this feature is rarely used, I can scrape off the implementation for now and focus on the `trafilatura` extraction. However, if we decide to keep it as a nice-to-have feature in this extractor I can readjust it accordingly.

I'll also cross-check why the pre-commit hooks have not done the proper formatting on my end.

Reply:
1. I definitely prefer the multi statement approach.
2. [Parsing only part of the document](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#improving-performance) may help?

>Parsing only part of a document won’t save you much time parsing the document, but it can save a lot of memory, and it’ll make searching the document much faster.
>